lms/djangoapps/experiments/tests/test_views.py

```diff
@@ -288,3 +293,13 @@ def test_permissions(self):
 
         response = self.client.delete(url)
         self.assertEqual(response.status_code, 403)
+
+class ExperimentUserMetaDataViewTests(APITestCase, ModuleStoreTestCase):
+
+    def test_UserMetaDataView_get(self):
+        """ Internal user_metadata view/API for use in Optimizely experiments """
+        user = UserFactory()
+        course = CourseFactory.create(start=now() - timedelta(days=30))
+        call_args = [user.username, course.id]
+        response = self.client.get(reverse('api_experiments:user_metadata', args=call_args))
+        self.assertEqual(response.status_code, 200)
```

Comment:
Kept this minimal too since we generally 'don't test for experiments' and this API is only for our internal use from optimizely to work around courseware MFE not having user_metadata.

Reply:
One thing to consider is how often we're going to be using the endpoint. Do we want to reuse it for more experiments than just the value prop experiment? How much do we care about getting this data for every single use case/configuration/special cases?

Reply:
Yes the idea is that we'll need this for many future experiments since the new courseware MFE doesn't have this info anymore (hence the endpoint rather than a cookie hack).  The endpoint will be called by Optimizely from these pages, where it has a user/course object, so it won't get back a 404 for example, and it won't be mis-forming its request once we have it set up. The only other special case I'd anticipate is some server issue/timeout in completing the request, but if that's happening then the Optimizely experiment is sort of the least of our worries. That's my thinking... is there anything I'm missing?

Reply:
The biggest one I can think of is access control - e.g. making sure that whoever is allowed access to this endpoint can see the data, and whoever doesn't have access to this endpoint can't see the data.

Another set of things you could test would be what data you get if the username is not passed in, the course_id is not passed in, and if both are not passed in, so you can confirm the behavior of when input data is missing.

Reply:
Yeah I originally had a test for missing arguments (it 404s), but decided it would be sort of imaginary/meaningless if all requests are coming from Optimizely/us with both pieces of information. I can add it back for completion's sake. 

Regarding access: I've added the same authentication_classes and permission_classes uses in another experiment class, but I'll have to look into how this works with Optimizely.

Reply:
Update: the original permission restricts results to an edx user. I've made a new permission now to restrict it to only giving a response for THAT SAME user (as the one who's logged in and making the request), will push as soon as I rework the tests accordingly.

Reply:
(this is also covered in thread below, so I'll resolve this one now that the missing-argument handling has been added- thanks!)